cmd/orch-tester/orch_tester.go

```diff
@@ -549,3 +599,209 @@ func (s *streamerClient) getOrchestrators() ([]*orch, error) {
+	randProfile := s.profiles[rand.Int()%len(s.profiles)]
+	//stream/manifestID/profile/SeqNo.ts
+	sourceUrl := fmt.Sprintf("%v/stream/%v_0_0/source/%v.ts", s.broadcaster, mid, randSeqNo)
+	renditionUrl := fmt.Sprintf("%v/stream/%v_0_0/%v/%v.ts", s.broadcaster, mid, randProfile, randSeqNo)
+	var (
+		rendition []byte
+		source    []byte
+		err       error
+	)
+	tick := time.NewTicker(2 * time.Second)
+	for {
+		select {
+		case <-ctx.Done():
+			glog.Info("random sampler timed out")
+			// If we haven't found the segment try to sample from whatever playlist we can
+			rendition, source, renditionUrl, err = s.sampleFromPlaylist(mid)
+			if err != nil {
+				glog.Error(err)
+			}
+			break
+		case <-tick.C:
+			glog.V(model.DEBUG).Infof("Trying to download %v", renditionUrl)
+			rendition, err = s.downloadSegment(renditionUrl)
+			if err != nil {
+				glog.V(model.DEBUG).Info(err)
+				continue
+			}
+			source, err = s.downloadSegment(sourceUrl)
+			if err != nil {
+				glog.V(model.DEBUG).Info(err)
+				continue
+			}
+			break
+		}
+		break
+	}
+
+	if len(rendition) == 0 || len(source) == 0 {
+		return fmt.Errorf("no segments found")
+	}
+
+	y, m, d := start.Date()
+	baseFileName := fmt.Sprintf("%v-%v%v", y, m.String(), d)
+	u := strings.Split(renditionUrl, "/")
+	u = u[len(u)-2:]
+	rendF := fmt.Sprintf("%v/%v/%v-%v", baseFileName, orch, mid, strings.Join(u, "-"))
+	srcF := fmt.Sprintf("%v/%v/%v-source-%v", baseFileName, orch, mid, u[len(u)-1])
+	src, _, err := testers.SaveToExternalStorage(srcF, source)
+	if err != nil {
+		glog.Error(err)
+		return err
+	}
+	glog.Infof("Wrote segment to storage file=%v", src)
+
+	rend, _, err := testers.SaveToExternalStorage(rendF, rendition)
+	if err != nil {
+		glog.Error(err)
+		return err
+	}
+	glog.Infof("Wrote segment to storage url=%v", rend)
+	return nil
+}
+
+func (s *streamerClient) sampleFromPlaylist(mid string) (source, rendition []byte, url string, err error) {
+	glog.Info("Sample not found, sampling from playlist")
+	mpl, err := s.downloadMasterPlaylist(mid)
+	if err != nil {
+		return nil, nil, "", err
+	}
+	r := rand.Int()
+	sourceURI := fmt.Sprintf("%v_0_0/source.m3u8", mid)
+	sourcePl, err := s.downloadMediaPlaylist(sourceURI)
+	if err != nil {
+		return nil, nil, "", err
+	}
+	renditionPl, err := s.downloadMediaPlaylist(mpl.Variants[rand.Int()%len(mpl.Variants)].URI)
+	if err != nil {
+		return nil, nil, "", err
+	}
+
+	if len(renditionPl.Segments) == 0 {
+		return nil, nil, "", fmt.Errorf("no transcoded segments")
+	}
+
+	var rendSegs []*m3u8.MediaSegment
+	var sourceSegs []*m3u8.MediaSegment
+	for _, v := range renditionPl.Segments {
+		if v != nil {
+			rendSegs = append(rendSegs, v)
+		}
+	}
+	for _, v := range sourcePl.Segments {
+		if v != nil {
+			sourceSegs = append(sourceSegs, v)
+		}
+	}
+	if len(rendSegs) == 0 || len(sourceSegs) == 0 {
+		return nil, nil, "", fmt.Errorf("no segments found")
+	}
+
+	seqNo := r % int(math.Min(float64(len(rendSegs)), float64(len(sourceSegs))))
```

Comment:
Instead of constructing the extra `rendSegs` and `sourceSegs` slices, you should be able to just do:

```
segIdx := r % int(math.Min(float64(len(renditionPl.Segments)), float64(len(sourcePl.Segments)))
seqNo := sourcePl.Segments[segIdx].SeqId // Assumes that segIdx points to the same seqNo in the slice of sources and renditions
```

Reply:
The loops filter out the nil values